source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,538 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file azure_iot_hub_client.c
+ * @brief -------.
+ */
+
+#include "azure_iot_hub_client.h"
+#include "azure_iot_mqtt.h"
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+#include "azure/az_iot.h"
+#include "azure/core/az_version.h"
+
+#ifndef azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC
+    #define azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC    azureiotDEFAULT_TOKEN_TIMEOUT_IN_SEC
+#endif /* azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC */
+
+#ifndef azureiothubKEEP_ALIVE_TIMEOUT_SECONDS
+    #define azureiothubKEEP_ALIVE_TIMEOUT_SECONDS    azureiotKEEP_ALIVE_TIMEOUT_SECONDS
+#endif /* azureiothubKEEP_ALIVE_TIMEOUT_SECONDS */
+
+#ifndef azureiothubSUBACK_WAIT_INTERVAL_MS
+    #define azureiothubSUBACK_WAIT_INTERVAL_MS    azureiotSUBACK_WAIT_INTERVAL_MS
+#endif /* azureiothubSUBACK_WAIT_INTERVAL_MS */
+
+#ifndef azureiothubUSER_AGENT
+    #define azureiothubUSER_AGENT    "DeviceClientType=c%2F" AZ_SDK_VERSION_STRING "%28FreeRTOS%29"
+#endif /* azureiothubUSER_AGENT */
+
+/*
+ * Topic subscribe state
+ */
+#define azureiothubTOPIC_SUBSCRIBE_STATE_NONE       0
+#define azureiothubTOPIC_SUBSCRIBE_STATE_SUB        1
+#define azureiothubTOPIC_SUBSCRIBE_STATE_SUBACK     2
+
+/*-----------------------------------------------------------*/
+
+static void prvMQTTProcessIncomingPublish( AzureIoTHubClientHandle_t xAzureIoTHubClientHandle,
+                                           AzureIoTMQTTPublishInfo_t * pxPublishInfo )
+{
+    uint32_t index;
+
+    configASSERT( pxPublishInfo != NULL );
+    AzureIoTHubClientReceiveContext_t * context;
+
+    for( index = 0; index < ( sizeof( xAzureIoTHubClientHandle->_internal.xReceiveContext ) / sizeof( xAzureIoTHubClientHandle->_internal.xReceiveContext[ 0 ] ) ); index++ )
+    {
+        context = &xAzureIoTHubClientHandle->_internal.xReceiveContext[ index ];
+
+        if( ( context->_internal.pxProcessFunction != NULL ) &&
+            ( context->_internal.pxProcessFunction( context,
+                                                   xAzureIoTHubClientHandle,
+                                                   ( void * ) pxPublishInfo ) == eAzureIoTHubClientSuccess ) )
+        {
+            break;
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static void prvMQTTProcessResponse( AzureIoTHubClientHandle_t xAzureIoTHubClientHandle,
+                                    AzureIoTMQTTPacketInfo_t * pxIncomingPacket,
+                                    uint16_t usPacketId )
+{
+    uint32_t index;
+    AzureIoTHubClientReceiveContext_t * context;
+
+    if( ( pxIncomingPacket->type & 0xF0U ) == AZURE_IOT_MQTT_PACKET_TYPE_SUBACK )
```

Comment:
How do we process PUBACK's and pass up to application?  I get it may not yet be plumbed yet, but even looking at the public header of this it's not clear how this is supposed ot even work in 1st place

From my memory and a very quick review of CoreMQTT, I don't think the publish call blocks until the PUBACK is received but just till bytes get sent across.  It seems -again quick - in CoreMQTT than in `handlePublishAcks` is where this is plumbed up to app but that `handlePublishAcks` is itself in the receive loop someplace.